File: LeetCode/src/BackTracking/M698_划分为k个相等的子集.java
package BackTracking;

import java.util.Arrays;

/**
 * @author Hyperspace
 * @date 2024/08/25
 * @file M698_划分为k个相等的子集.java
 * <p>
 * 思路
 * 回溯
 * <p>
 * 有一个很重要的优化
 * idx > 0 && pos[idx] == pos[idx - 1]
 * 若当前元素和与上一次相同，其选法没有区别，可剪枝
 */
public class M698_划分为k个相等的子集 {
    int[] nums, pos;
    int target;
    boolean ans;

    public boolean canPartitionKSubsets(int[] nums, int k) {
        this.nums = nums;
        pos = new int[k];
        int n = nums.length;
        int sum = Arrays.stream(nums).sum();
        if (sum % k != 0) {
            return false;
        }
        target = sum / k;
        int t;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (nums[i] >= nums[j]) {
                    t = nums[i];
                    nums[i] = nums[j];
                    nums[j] = t;
                }
            }
        }
        backtrack(0);

        return ans;
    }

    private void backtrack(int i) {
        if (i == nums.length) {
            ans = true;
            return;
        }
        for (int idx = 0; idx < pos.length; idx++) {
            if (pos[idx] + nums[i] > target || (idx > 0 && pos[idx] == pos[idx - 1])) {
                continue;
            }
            pos[idx] += nums[i];
            backtrack(i + 1);
            pos[idx] -= nums[i];
        }
    }
}

File: GolangSolutions/BackTracking/M698_划分为k个相等的子集.go
package BackTracking

import "sort"

/*
@author Hyperspace
@date 2024/08/25
@file M698_划分为k个相等的子集.go
*/
func canPartitionKSubsets(nums []int, k int) bool {
	sum := 0
	for _, v := range nums {
		sum += v
	}
	if sum%k != 0 {
		return false
	}
	target := sum / k
	pos := make([]int, k)
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] > nums[j]
	})

	return backtrack(pos, nums, k, 0, target)
}

func backtrack(pos, nums []int, k, i, target int) bool {
	if i == len(nums) {
		return true
	}
	for idx := range pos {
		if pos[idx]+nums[i] > target || (idx > 0 && pos[idx] == pos[idx-1]) {
			continue
		}
		pos[idx] += nums[i]
		if backtrack(pos, nums, k, i+1, target) {
			return true
		}
		pos[idx] -= nums[i]
	}

	return false
}
